### PR TITLE
[Proxy] feat: implement relayer proxy full workflow

### PR DESCRIPTION
**Pull request by red-0ne**
### Human Summary

Implemented the needed parts of the `RelayerProxy` flow which involves:
- Building the `RelayRequest` from the `http.Request`
- Verifying the `RelayRequest` signature and session
- Forwarding the request to the native service
- Building the `RelayReponse` and signing it
- Replying to the requester
- Notifying the serverdRelays observable
- Replying appropriately if any error happens

It assumes the availability `BlockClient` and `Session` query client with the appropriate `SessionHeader` fields (available when PRs #78 and #65 get merged)

### Summary generated by Reviewpad on 08 Nov 23 22:31 UTC
This pull request includes various changes across multiple files in the codebase. Here's a summary of the changes:
1. In the file "file_diff.go":
   - Updated imports to include the packages "github.com/pokt-network/poktroll/pkg/relayer" and "github.com/pokt-network/poktroll/x/supplier/types".
   - Added a new variable "rp.supplierAddress" and modified the types of two variables: "serviceEndpoints" and "server" to use types from the "relayer" package.

2. In the file "relay_signer.go":
   - Introduced a new file that contains a method to sign the hash of the RelayResponse.
   - The method relies on the keyring and keyName to sign the payload and returns the signature.
   - There are TODO_TECHDEBT comments suggesting improvements for the method.

3. In the file "interface.go":
   - Added new imports and defined interfaces for managing relay servers and relay sessions.
   - These interfaces provide methods for starting and stopping relay servers, retrieving served relays, and managing SMSTs.
   - Removed unused methods and modified existing type declarations.

4. In the file "error_reply.go":
   - Introduced a new file that contains a method to generate and write JSONRPC error responses.
   - The method builds an error response based on the input error and writes it to the HTTP ResponseWriter.
   - There are TODO_TECHDEBT comments suggesting improvements for the method.

5. In the file "other_file_diff.go":
   - Added imports for various packages.
   - Updated type declarations and added new methods.
   - Removed an unused interface.

6. In the file "interface.go":
   - This file has been deleted. It contained interfaces for managing relay servers and relaying requests, which are no longer needed.

7. In the file "proxy_test.go":
   - Introduced a new file that contains tests for the "relayerProxy" and its jsonRPC component.
   - There is a TODO comment suggesting adding more tests.

8. In the file "relay.proto":
   - Made changes to import statements and uncommented a message field in the RelayRequestMetadata message and the RelayResponseMetadata message.

9. In the file "jsonrpc.go":
   - Made changes related to implementing a JSON-RPC server for relaying requests.
   - Added imports, changed type assertions, and added new methods.
   - Modified the logic for serving relay requests and sending responses.

10. In the file "relay_verifier.go":
    - Introduced a new file that contains a method to verify relay requests.
    - The method checks the relay request signature, session validity, and whether the relayer proxy can serve the request.
    - TODO comments suggest areas for further investigation.

11. In the file "errors.go":
    - Made changes to define new error variables for specific error scenarios.

12. In the file "new_file.go":
    - Added a new file that contains functions for building RelayRequest and RelayResponse objects from HTTP requests and responses.
    - The functions handle marshaling and unmarshaling of JSON data.

Please let me know if you need more information.

## Issue

- #14 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/proxy/jsonrpc.go [hunk 1]
```diff
@@ -78,4 +81,90 @@ func (j *jsonRPCServer) ServiceId() string {
 // when jsonRPCServer is used as an http.Handler with an http.Server.
 // (see https://pkg.go.dev/net/http#Handler)
 func (j *jsonRPCServer) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	ctx := request.Context()
+	// Relay the request to the native service and build the response that will be sent back to the client.
```
Comment (Olshansk) [inline pkg/relayer/proxy/jsonrpc.go:85]: I want us to avoid using the word "native". Consider "supported", "backing", "staked for", etc...

File: pkg/relayer/proxy/jsonrpc.go [hunk 2]
```diff
@@ -78,4 +81,90 @@ func (j *jsonRPCServer) ServiceId() string {
 // when jsonRPCServer is used as an http.Handler with an http.Server.
 // (see https://pkg.go.dev/net/http#Handler)
 func (j *jsonRPCServer) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
```
Comment (Olshansk) [inline pkg/relayer/proxy/jsonrpc.go:83]: OPTIONAL NIT: Can we rename the receiver from `j` to `jsrv` to it's easy to read as `json service`

File: pkg/relayer/proxy/jsonrpc.go [hunk 3]
```diff
@@ -78,4 +81,90 @@ func (j *jsonRPCServer) ServiceId() string {
...
+		j.replyWithError(writer, err)
+		return
+	}
+
+	// Send the relay response to the client.
+	if err := j.sendRelayResponse(relay.Res, writer); err != nil {
+		j.replyWithError(writer, err)
+		return
+	}
+
```
Comment (Olshansk) [inline pkg/relayer/proxy/jsonrpc.go:98]: Please add some logging on errors & on success.

E.g. `Successfully relayed response`

File: pkg/relayer/proxy/jsonrpc.go [hunk 4]
```diff
@@ -78,4 +81,90 @@ func (j *jsonRPCServer) ServiceId() string {
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify the relay request signature and session.
+	if err := j.relayerProxy.VerifyRelayRequest(ctx, relayRequest, j.serviceId); err != nil {
+		return nil, err
+	}
+
+	// Get the relayRequest payload's ReadCloser to add it to the http.Request
```
Comment (Olshansk) [inline pkg/relayer/proxy/jsonrpc.go:116]: What's `ReadCloser` #PUC

File: pkg/relayer/proxy/jsonrpc.go [hunk 5]
```diff
@@ -78,4 +81,90 @@ func (j *jsonRPCServer) ServiceId() string {
...
+		j.replyWithError(writer, err)
+		return
+	}
+
+	// Emit the relay to the servedRelays observable.
+	j.servedRelaysProducer <- relay
+}
+
+// serveHTTP holds the underlying logic of ServeHTTP.
+func (j *jsonRPCServer) serveHTTP(ctx context.Context, request *http.Request) (*types.Relay, error) {
```
Comment (Olshansk) [inline pkg/relayer/proxy/jsonrpc.go:104]: Very nice and easy to  understand function!

File: pkg/relayer/proxy/jsonrpc.go [hunk 6]
```diff
@@ -78,4 +81,90 @@ func (j *jsonRPCServer) ServiceId() string {
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.Relay{Req: relayRequest, Res: relayResponse}, nil
+}
+
+// sendRelayResponse marshals the relay response and sends it to the client.
+func (j *jsonRPCServer) sendRelayResponse(relayResponse *types.RelayResponse, writer http.ResponseWriter) error {
+	relayResposeBz, err := relayResponse.Marshal()
```
Comment (Olshansk) [inline pkg/relayer/proxy/jsonrpc.go:159]: s/relayResposeBz/relayResponseBz

File: pkg/relayer/proxy/jsonrpc.go [hunk 7]
```diff
@@ -78,4 +81,90 @@ func (j *jsonRPCServer) ServiceId() string {
...
+	if err != nil {
+		return err
+	}
+
+	_, err = writer.Write(relayResposeBz)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (Olshansk) [inline pkg/relayer/proxy/jsonrpc.go:169]:
```suggestion
	return err
```

File: pkg/relayer/proxy/relay_builders.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+// If the response has a non-nil body, it will be parsed as a JSONRPCResponsePayload.
+func (j *jsonRPCServer) newRelayResponse(
+	response *http.Response,
+	sessionHeader *sessiontypes.SessionHeader,
+) (*types.RelayResponse, error) {
+	relayResponse := &types.RelayResponse{
+		Meta: &types.RelayResponseMetadata{SessionHeader: sessionHeader},
+	}
+
+	if response.Body != nil {
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_builders.go:37]: If it's nil, should we just return an error and exit quickly?

File: pkg/relayer/proxy/relay_verifier.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+// VerifyRelayRequest is a shared method used by RelayServers to check the relay request signature and session validity.
+func (rp *relayerProxy) VerifyRelayRequest(
+	ctx context.Context,
+	relayRequest *types.RelayRequest,
+	serviceId serviceId,
+) error {
+	// Query for the application account to get the application's public key to verify the relay request signature.
+	applicationAddress := relayRequest.Meta.SessionHeader.ApplicationAddress
+	accountQuery := &accounttypes.QueryAccountRequest{Address: applicationAddress}
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_verifier.go:21]:
```suggestion
	accQueryReq := &accounttypes.QueryAccountRequest{Address: applicationAddress}
```

File: pkg/relayer/proxy/relay_verifier.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	accountResponse, err := rp.accountsQuerier.Account(ctx, accountQuery)
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_verifier.go:22]:
```suggestion
	accQueryResponse, err := rp.accountsQuerier.Account(ctx, accountQuery)
```

File: pkg/relayer/proxy/relay_verifier.go [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+	var payloadBz []byte
+	_, err = relayRequest.Payload.MarshalTo(payloadBz)
+	if err != nil {
+		return err
+	}
+	hash := crypto.Sha256(payloadBz)
+
+	// accountResponse.Account.Value is a protobuf Any type that should be unmarshaled into an AccountI interface.
+	// TODO_DISCUSS: Make sure this is the correct way to unmarshal an AccountI from an Any type.
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_verifier.go:35]: If you look inside `docs/static/openapi.yml`. They have this

```go
             Example 4: Pack and unpack a message in Go

                 foo := &pb.Foo{...}
                 any, err := anypb.New(foo)
                 if err != nil {
                   ...
                 }
                 ...
                 foo := &pb.Foo{}
                 if err := any.UnmarshalTo(foo); err != nil {
                   ...
                 }
```

tl;dr

1. Use the Account struct type
2. Use `any.UnmarshalTo`

File: pkg/relayer/proxy/relay_verifier.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+	sessionResponse, err := rp.sessionQuerier.GetSession(ctx, sessionQuery)
+	session := sessionResponse.Session
+
+	// Since the retrieved sessionId was in terms of:
+	// - the current block height (which is not provided by the relayRequest)
+	// - serviceId (which is not provided by the relayRequest)
+	// - applicationAddress (which is used to to verify the relayRequest signature)
+	// we can reduce the session validity check to checking if the retrieved session's sessionId
+	// matches the relayRequest sessionId.
+	if session.SessionId != relayRequest.Meta.SessionHeader.SessionId {
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_verifier.go:61]:
```suggestion
	// TODO_INVESTIGATE: Revisit the assumptions above at some point in the future, but good enough for now
	if session.SessionId != relayRequest.Meta.SessionHeader.SessionId {
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/relayer/proxy/relay_verifier.go (hunk 11 above)
Comment (red-0ne) [reply pkg/relayer/proxy/relay_verifier.go:35]: Couldn't make it work. I'll stick the current approach until we get tests or E2E

File: pkg/relayer/proxy/relay_builders.go (hunk 8 above)
Comment (red-0ne) [reply pkg/relayer/proxy/relay_builders.go:37]: It is safe to just remove the check as `io.ReadAll(nil)` would return an empty []byte

File: pkg/relayer/proxy/relay_verifier.go [hunk 13]
```diff
@@ -34,12 +35,12 @@ func (rp *relayerProxy) VerifyRelayRequest(
 	// accountResponse.Account.Value is a protobuf Any type that should be unmarshaled into an AccountI interface.
 	// TODO_DISCUSS: Make sure this is the correct way to unmarshal an AccountI from an Any type.
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_verifier.go:36]:
```suggestion
	// TODO_TECHDEBT: Make sure our `AccountI`/`any` unmarshalling is correct. See https://github.com/pokt-network/poktroll/pull/101/files/edbd628e9146e232ef58c71cfa8f4be2135cdb50..fbba10626df79f6bf6e2218513dfdeb40a629790#r1372464439

**Review by Olshansk — APPROVED** [review]
Just left one update in a TODO comment, but otherwise really clean and g2g!

File: pkg/relayer/proxy/relay_verifier.go (hunk 11 above)
Comment (h5law) [reply pkg/relayer/proxy/relay_verifier.go]: @Olshansk we should not use `anypb.New` instead use cosmos's own `Any` type as follows:
```go
codectypes.NewAnyWithValue(value)
```

**Comment by Olshansk** [comment]
@bryanchriswhite After you merge in #65, let's try to rebase & update this PR as it should be g2g.

**Review by Olshansk — CHANGES_REQUESTED** [review]
@red-0ne Please merge (and fix conflicts) with main, and will re-review then.

Please also add `*_test.go` files with a `TODO_TEST` comment where applicable explaining that we're biasing towards an `E2E relay` and that building test utilities for these off-chain components is a non-trivial amount of work.

File: pkg/relayer/proxy/interface.go
Comment (bryanchriswhite) [inline pkg/relayer/proxy/interface.go:1]: Similar feedback as https://github.com/pokt-network/poktroll/pull/104/files#r1384945249:

> I think we need to move this file up one directory so that is is in the `/pkg/relayer` package. This will mitigate future import cycles caused by the interface and its implementation sharing a package.
> 
> Additionally, this has the effect of consolidating all of the interfaces which are dependencies of relayer into a single place.

File: pkg/relayer/proxy/relay_signer.go [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package proxy
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_signer.go:1]: Can you leave a TODO in this file as well that the relayer proxy isn't the best place to have the signer.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Reviewed everything with the exception of `pkg/relayer/proxy/relay_verifier.go`. Going to check that out tomorrow. Great work!

File: pkg/relayer/proxy/jsonrpc.go [hunk 15]
```diff
@@ -57,29 +61,120 @@ func NewJSONRPCServer(
...
+// serveHTTP holds the underlying logic of ServeHTTP.
+func (jsrv *jsonRPCServer) serveHTTP(ctx context.Context, request *http.Request) (*types.Relay, error) {
+	// Extract the relay request from the request body.
+	relayRequest, err := jsrv.newRelayRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify the relay request signature and session.
+	// TODO_TECHDEBT: Currently, the relayer proxy is responsible for verifying
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/relayer/proxy/jsonrpc.go:120]: https://github.com/pokt-network/poktroll/issues/160

**Comment by red-0ne** [comment]
This PR should be merged after #150

**Comment by red-0ne** [comment]
@h5law , we switched from only signing the `RelayResponse` payload to signing the whole response (including session headers). Please make sure to reflect this in the `AppGate` response verification.

File: pkg/relayer/interface.go [hunk 16]
```diff
@@ -27,10 +27,21 @@ type RelayerProxy interface {
 
 	// VerifyRelayRequest is a shared method used by RelayServers to check the
 	// relay request signature and session validity.
-	VerifyRelayRequest(relayRequest *types.RelayRequest) (isValid bool, err error)
-
-	// SignRelayResponse is a shared method used by RelayServers to sign the relay response.
-	SignRelayResponse(relayResponse *types.RelayResponse) ([]byte, error)
+	// This method may mutate the relay request in the future, so the returned
```
Comment (Olshansk) [inline pkg/relayer/interface.go:30]: I think it's kind of confusing that:
1. We pass in a pointer to an object
2. That pointer may be mutated
3. We return it

I believe it should be one of:
1. Pass a pointer that may be mutated
2. Pass a value that's copied, mutated, and returned

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Comment by Olshansk** [comment]
@red-0ne Also, please update this PR w/ main

File: pkg/relayer/interface.go (hunk 16 above)
Comment (red-0ne) [reply pkg/relayer/interface.go:30]: I see... I'll revert to passing the pointer without returning and mutate it inside the function as I can't clearly foresee the impact of changing the whole workflow to not use pointers. We'll keep that question to when we start implementing the middleware capabilities.

Comment (red-0ne) [reply pkg/relayer/interface.go:30]: cc @bryanchriswhite

**Review by Olshansk — APPROVED** [review]
